### DISCUSSION TITLE
db.Start() should take context

### DISCUSSION THREAD

**Discussion by aathan** [olric-data/olric#176] [Ideas]
Shouldn't the db.Start() method take a ctx parameter so that it can observe a timeout for the start procedure? Otherwise, an app that depends on olric must block forever (as far as what's implied by the function signature, despite whether or not Olric's startup logic contains the possibility of a forever block).

EDIT: Having understood the interface a bit further, I guess db.Start() is better called db.Run() and is intended to be executed as a blocking goroutine. In which case, I might suggest that Start() should be a ctx/cancelable operation with timeout support that starts a goroutine for the service, so that embedded deployments can control startup failures / delays more effectively?

**Comment by buraksezer** [comment]
`Config` struct has a callback function called `Started`. 

```go
type Config struct {
	// Callback function. Olric calls this after
	// the server is ready to accept new connections.
	Started func()
}
```

Olric calls this function when everything is started successfully. `Start` has a complicated business logic. You may try to call `Start` function in a goroutine and call `Shutdown` in a different goroutine in the case of timeout. It should work.
